tempodb/encoding/vparquet/block_search.go

```diff
@@ -317,14 +422,121 @@ func rawToResults(ctx context.Context, pf *parquet.File, rgs []parquet.RowGroup,
+	rgs := pf.RowGroups()
+	makeIter := makeIterFunc(ctx, rgs, pf)
+
+	keyPred := pq.NewStringInPredicate([]string{tag})
+
+	iter := pq.NewJoinIterator(DefinitionLevelResourceAttrs, []pq.Iterator{
+		makeIter(FieldResourceAttrKey, keyPred, "keys"),
+		makeIter(FieldResourceAttrVal, nil, "values"),
+	}, nil)
+	for {
+		match := iter.Next()
+		if match == nil {
+			break
+		}
+		m := match.ToMap()
+		for _, s := range m["values"] {
+			cb(s.String())
+		}
+	}
+	iter.Close()
+
+	iter = pq.NewJoinIterator(DefinitionLevelResourceSpansILSSpan, []pq.Iterator{
+		makeIter(FieldSpanAttrKey, keyPred, "keys"),
+		makeIter(FieldSpanAttrVal, nil, "values"),
+	}, nil)
+	for {
+		match := iter.Next()
+		if match == nil {
+			break
+		}
+		m := match.ToMap()
+		for _, s := range m["values"] {
+			cb(s.String())
+		}
+	}
+	iter.Close()
+
+	return nil
+}
+
+// searchSpecialTagValues searches a parquet file for all values for the provided column. It first attempts
+// to only pull all values from the column's dictionary. If this fails it falls back to scanning the entire path.
+func searchSpecialTagValues(ctx context.Context, column string, pf *parquet.File, cb common.TagCallback) error {
+	idx, _ := pq.GetColumnIndexByPath(pf, column)
+	if idx == -1 {
+		return fmt.Errorf("column not found: %s", column)
+	}
+
+	// now search all row groups
+	rgs := pf.RowGroups()
+	dictionarySearch := true
+rowgroups:
+	for _, rg := range rgs {
+		cc := rg.ColumnChunks()[idx]
+
+		pgs := cc.Pages()
+		for {
+			pg, err := pgs.ReadPage()
+			if err == io.EOF || pg == nil {
+				break
+			}
+			if err != nil {
+				return err
+			}
+
+			// if this column has a dictionary we are in luck!
+			dict := pg.Dictionary()
+			if dict != nil {
+				for i := 0; i < dict.Len(); i++ {
+					s := string(dict.Index(int32(i)).ByteArray())
+					cb(s)
+				}
+				break
+			}
+
+			// if this column doesn't have a dictionary we need to bite the bullet and iterate every page, bail!
+			dictionarySearch = false
+			continue rowgroups
+		}
+	}
+
+	if !dictionarySearch {
+		makeIter := makeIterFunc(ctx, rgs, pf)
+
+		iter := makeIter(column, nil, "values")
+		for {
+			match := iter.Next()
+			if match == nil {
+				break
+			}
+			m := match.ToMap()
+			for _, s := range m["values"] {
+				cb(s.String())
+			}
+		}
+		iter.Close()
+	}
+
+	return nil
 }
 
-func (p *parquetSearchMetrics) ToProto() *tempopb.SearchMetrics {
-	return &tempopb.SearchMetrics{
-		//InspectedTraces: uint32(p.pTraceID.InspectedValues.Load()),
+func makeIterFunc(ctx context.Context, rgs []parquet.RowGroup, pf *parquet.File) func(name string, predicate pq.Predicate, selectAs string) pq.Iterator {
+	return func(name string, predicate pq.Predicate, selectAs string) pq.Iterator {
+		index, _ := pq.GetColumnIndexByPath(pf, name)
+		if index == -1 {
+			// TODO - don't panic, error instead
+			panic("column not found in parquet file:" + name)
+		}
+		if selectAs == "" {
```

Comment:
This change increases the amount of cpu work done.  The iterator uses an empty selectAs to avoid saving the values in the IteratorResults where we filter the column but don't need to return the values upstream.

Reply:
nice catch. removed and `rawToResults` has been updated to explicitly name its columns.